### PR TITLE
store connection.rawConnection.params.rawBody on web connections to keep around raw POST body

### PR DESCRIPTION
**Pull request by evantahler**
solves https://github.com/actionhero/actionhero/issues/1081

### CODE REVIEW

File: servers/web.js [hunk 1]
```diff
@@ -505,12 +505,17 @@ const initialize = function (api, options, next) {
...
+        let rawBody = []
+        connection.rawConnection.req.on('data', (chunk) => { rawBody.push(chunk) })
+
         connection.rawConnection.form.parse(connection.rawConnection.req, (error, fields, files) => {
           if (error) {
             server.log('error processing form: ' + String(error), 'error')
             connection.error = new Error('There was an error processing this form.')
           } else {
             connection.rawConnection.params.body = fields
+            connection.rawConnection.params.rawBody = Buffer.concat(rawBody).toString().substring(0, api.config.servers.web.formOptions.maxFieldsSize)
```
Comment (evantahler) [inline servers/web.js]: normally, we would need to wait until `connection.rawConnection.req.on('end')`, but formidable does the same thing, we can wait within it's callback

**Comment by evantahler** [comment]
@crrobinson14 @genexp This is likely to increase the memory footprint of every connection... good idea or bad idea?

**Comment by evantahler** [comment]
I guess the thing to do here would be to make this optional!

**Comment by crrobinson14** [comment]
@evantahler Aside from the length checking ability it adds, why jump through the Buffer/toString hoops to copy rawBody out? It seems like it's extra processing that may not be necessary?

Two possible options that might be an enhancement?
1. Instead of making this a global setting, make it metadata in the action. I think it's likely that only one or two API calls in most stacks will need this, so most devs probably don't need to turn it on for every call, just weird stuff like Stripe (which signs the body with a header field so you need to check the raw body).
2. Just store a ref to rawBody, but add a connection class method to obtain it, such as `.getRawBody()`. That could do the buffer/string/length limit dance only where required. This doesn't reduce the memory footprint (because a ref is saved for rawBody, of course), but defers it until other things have done. What if input validation fails, or other middleware aborts the connection? The raw body processing will have been wasted - making it a great DDOS make-work vector...

**Comment by evantahler** [comment]
Excellent ideas! I'll work some of them up now.  You hit the nail on the head: a lot of this is dangerous (memory, DDOS, etc).

**Comment by evantahler** [comment]
@crrobinson14 I think we would still need to store the body-buffer somewhere, and since we want it destroyed along with the connection, it seems to make sense to append it there.  Regarding the buffer-to-string, you are right! might as well let the consuming action handle that.

**Comment by evantahler** [comment]
regarding action metadata, I really don't like action-metadata which is only useful for one type of connection.  We only have one of those and it tears me up inside :D

More importantly, the connection is fully parsed (and params extracted) before the action is even determined.  We can't wait until then to know if we need to keep the body around.

**Comment by krlicmuhamed** [comment]
Can I read a request body Stream in some action instead of It just storing complete body data?

**Comment by evantahler** [comment]
The problem is that the stream is complete by the time the action is started.  It needs to work this way because we need all FROM options/uploads, from the body stream, to be available (and added to `params`) for the action.

**Comment by krlicmuhamed** [comment]
Maybe we can add another function in `exports.action` that runs before body stream starts?
That way we can have _some_ shared memory with server code. 
Something like this:
```
var rawBody;
exports.action = {
 name: 'test',
 initialize: function(shared){
  shared.events.on('rawbodydata', function(raw){
   ...
   rawBody = raw;
  });
 }
 run: function (api, data, next) {}
}
```
This is the `connection.rawConnection.params.rawBody` done in a different way though, except you gain more flexibility :)

**Comment by evantahler** [comment]
I really don't like the implementation of this pull request... so send you idea as a pull request!

**Pull request merged**